### ISSUE TITLE
[Bug]: optimizer.zero_grad() called after optimizer.step() in compression distillation — incorrect gradient accumulation

### ISSUE THREAD

**Issue by Prathvikmehra** [Adit-Jain-srm/NightmareNet#483]
## Problem / Motivation

In 
ightmarenet/compression/distillation.py:162-164, the training step order is:

`python
loss.backward()     # computes gradients (ADDING to existing .grad)
optimizer.step()    # applies gradients (which include stale gradients from previous step!)
optimizer.zero_grad()  # zeroes gradients TOO LATE
`

Standard PyTorch practice is zero_grad() → ackward() → step(). Because zero_grad() is called *after* step(), gradients from iteration N are **never cleared** before iteration N+1's ackward(). This means ackward() on iteration N+1 **accumulates** into the un-zeroed gradients from iteration N, and step() on iteration N+1 uses the sum of N and N+1 gradients.

This effectively implements **unintentional gradient accumulation** where the accumulation window grows unbounded (the window for step N includes all prior steps' gradients since the last zero_grad() — which never happens between steps).

### Impact
Model training in the compression/distillation phase has incorrect gradient values, leading to suboptimal or diverging weight updates. This affects every pipeline run that uses compression.

### Affected Files
- 
ightmarenet/compression/distillation.py:162-164

## Proposed Solution
Swap lines 162 and 164 so the order is:
`python
optimizer.zero_grad()  # NEW: clear gradients from previous step
loss.backward()        # compute fresh gradients
optimizer.step()       # apply fresh gradients
`

## Acceptance Criteria
- [ ] zero_grad() executes before ackward() in the distillation training step
- [ ] Training loss converges as expected (compare with baseline)
- [ ] All existing compression/distillation tests pass

## Scope / Difficulty Estimate
Easy — 15 min. Single-line swap. However, retraining validation is needed to confirm loss behavior changes as expected.

**Comment by Adit-Jain-srm** [comment]
Closing. This is not a bug.

The existing pattern loss.backward() -> optimizer.step() -> optimizer.zero_grad() is standard PyTorch. zero_grad() after step() clears gradients for the NEXT iteration — it does NOT cause gradient accumulation across steps.

Both "zero before backward" and "zero after step" are equivalent for single-loss-per-step training. The current code is correct. The issue this PR claims to fix does not exist.

Reference: [PyTorch docs](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html) — zero_grad() placement is a style choice when not accumulating gradients intentionally.
